designdocs/recovery_system.md

```diff
@@ -0,0 +1,505 @@
+
+#### 4. New Key Generation
+
+Once validation passes, the provider generates a new Control Key for the MSA. There are two approaches:
+
+- User-specified key: If the user is somewhat technical, they might generate a new key pair on their own device (e.g. via a wallet app) and only give the public key to the provider. This is ideal for security, as the private key never leaves the user. The recovery provider would ask “Please paste a new public key to add to your account.” However, this requires the user to know how to create a key pair, which may not be user-friendly in an emergency.
+- Provider-generated key: The recovery provider can create a new key pair on behalf of the user. For instance, it generates a new Ed25519, Sr25519, or Ethereum key (compatible with Frequency). The provider will then securely transmit the private key to the user after the on-chain update. This approach is simpler for the user but requires trust that the provider deletes its copy. In practice the provider could generate a mnemonic and show it to the user over a secure TLS connection for them to copy, minimizing its own storage of the key.
+
+This design recommends giving the user the option, but in either case, a public key for the new control key is decided at this step.
+
+#### 5. On-Chain Recovery Execution
+
+The recovery provider now submits a special Recovery extrinsic to the Frequency blockchain. This transaction does two main things atomically:
+
+ 1. Invalidate Old Recovery Key: It marks the stored recovery key for `target_msa` as used. This can be done by clearing the hash or flagging it so it cannot be reused. (For example, the extrinsic could set the recovery hash storage to 0 or remove that entry for the MSA.)
+ 2. Add New Control Key: It adds the new public key to the list of control keys for `target_msa`. Frequency allows multiple control keys per MSA, so the chain will allow a recovery provider to add a new public key. The user must delegate this authority to the recovery provider. In this hash design, an elegant solution is: the extrinsic call could include the plaintext recovery code (or its hash) as a parameter; the runtime hashes it and verifies it matches the stored hash for that MSA. This way, the provider proves on-chain that it indeed has the secret. To avoid exposing the secret in plaintext on-chain, the extrinsic could take H’ (the hash) as input instead. But since H’ is stored and public, an attacker provider could try to reuse it; thus requiring the actual secret or a signature is safer. Another approach is having the provider sign the transaction with its own key (proving it’s authorized) and relying on off-chain enforcement of email verification (since governance oversight exists). For maximal security, we can do: recoverAccount(msa_id, recovery_code, new_pubkey) → runtime computes hash of (email||secret) and checks match & provider authorization. This ensures only a correct recovery_code unlocks the addition.
```

Comment:
Per note above, I prefer submitting `H'` with a provider signature (we need proof that the provider is an authorized recovery provider); the chain would store `H(H')`.